### DISCUSSION TITLE
Dialect Development: Debug reusable compliance tests with VSCode

### DISCUSSION THREAD

**Discussion by susodapop** [sqlalchemy/sqlalchemy#10940] [Usage Questions]
Please let me know if I'm posting in the wrong spot.

I maintain a dialect for connecting to Databricks with sqlalchemy. I'm pleased to use the reusable dialect compliance test suite and have found it very useful for understanding the library's inner-workings. But right now, I'm only able to run the tests from a command line with `pytest -m pytest path/to/test_suite.py --dburi "databricks://..."`.

I'd like to be able to run the tests in VSCode's visual debugger. Is there a guide anywhere for how to set this up?

I can make VSCode discover the tests. But when I click to run one, it fails because the sqlalchemy pytest plugin [renames the test-cases](https://github.com/zzzeek/sqlalchemy/blob/582058ea54ba9445c7bb0ffd94690f77ce55a5ba/lib/sqlalchemy/testing/plugin/plugin_base.py#L545-L566) in such a way that they cannot be invoked. For example: `DateTimeTest` becomes `DateTimeTest_+databricks`. But if I run `python -m pytest test_suite.py::DateTimeTest_+databricks` this fails.

As a debug step, I've tried manually patching `plugin_base.py` to not re-write these test case names. But doing breaks the dialect-lookup behaviour of the test plugin, so that running `python -m pytest test_suite.py::DateTimeTest` tries to use the sqlite dialect rather than my custom dialect. Any pointers?

**Comment by zzzeek** [comment]
` test_suite.py::DateTimeTest_+databricks` does not sound like the correct name so the first step would be to be able to run individual tests from a regular console command line.   If you run pytest for test_suite.py overall, it will print out the full names.    the names should have the dialect name and driver name, followed by the server version number reported by the database, like this:

    test/dialect/test_suite.py::ComponentReflectionTestExtra_sqlite+pysqlite_3_42_0::test_reflect_expression_based_indexes

you can run a name like that from pytest directly.    as far as how vscode integrates with that, while I do use vscode for everything I dont use it to actually run pytest, so i dont know how that works or what it expects.

**Reply by susodapop** [reply]
Thanks @zzzeek for that context.

>  test_suite.py::DateTimeTest_+databricks does not sound like the correct name

This helped me find a gap in my dialect. Once I added a `.driver` attribute and a `_get_server_version_info()` method to my `Dialect`, running `pytest test_suite.py --collect-only` spits out more sensible names. 

```
<Class ComponentReflectionTest_databricks+databricks_0_0_1>
      <Function test_autoincrement_col>
      <Function test_comments_unicode>
      <Function test_comments_unicode_full>
      <Function test_dialect_initialize>
```

> so the first step would be to be able to run individual tests from a regular console command line.

I'm with you so far. Interestingly, I've been able to do this already. I can even select individual classes and test methods within those classes, like so:

```
pytest test_suite.py::ComponentReflectionTest::test_get_multi_table_comment
```

> you can run a name like that from pytest directly. 

This is where things break down. If I run the following it fails, even though this exact name is printed when `--collect-only` is used:

```
pytest test_suite.py::ComponentReflectionTest_databricks+databricks_0_0_1::test_get_multi_table_comment
```

From what I can tell, the inclusion of the `name`, `driver`, and `server_version` in the test case name is meant for running a matrix of different dialects, drivers, and server versions. The pytest plugin includes a `before_test()` method that fetches the class's `_sa_orig_cls_name` and runs that instead. I tried to `breakpoint()` into `before_test()` to watch this happen but the `not found` exception is raised before that code even runs.

Should pytest be able to run the test case including the `dialect/driver/version` tag?

**Reply by zzzeek** [reply]
Oh I misspoke a bit.   the dialect/server name thing prints out on the screen but to send the test name in from pytest, you omit that part:

     pytest test/dialect/test_suite.py::ComponentReflectionTest::test_get_multi_table_comment

then when it runs the test name is dynamically parameterized with the dialect etc. and in some cases the pytest parametrize also like this:

```
test/dialect/test_suite.py::ComponentReflectionTest_sqlite+pysqlite_3_42_0::test_get_multi_table_comment[False-ObjectKind.ANY-ObjectScope.ANY-<lambda>-_exclusions_01] SKIPPED [  1%]
test/dialect/test_suite.py::ComponentReflectionTest_sqlite+pysqlite_3_42_0::test_get_multi_table_comment[False-ObjectKind.ANY-ObjectScope.ANY-None-_exclusions_00] SKIPPED [  2%]
test/dialect/test_suite.py::ComponentReflectionTest_sqlite+pysqlite_3_42_0::test_get_multi_table_comment[False-ObjectKind.ANY-ObjectScope.DEFAULT-<lambda>-_exclusions_01] SKIPPED [  3%]
test/dialect/test_suite.py::ComponentReflectionTest_sqlite+pysqlite_3_42_0::test_get_multi_table_comment[False-ObjectKind.ANY-ObjectScope.DEFAULT-None-_exclusions_00] SKIPPED [  4%]
...

```

**Reply by CaselIT** [reply]
@susodapop I also don't use the vscode test plug-in much, but I did notice that it had issue running selected tests. I don't remember if I solved it in some way of not, will check once I'm in the other PC

If you are familiar with pytest plug-ins and want to check if there is a better way / location to do the parametrization that's currently performed feel free to experiment.

Maybe attaching to the class a parametrize or parametrized fixture may be an option here? It would mess a bit with the current naming scheme since if would probably result in something like 
```
test/dialect/test_suite.py::ComponentReflectionTest::test_get_multi_table_comment[sqlite+pysqlite_3_42_0-False-ObjectKind.ANY-ObjectScope.ANY-<lambda>-_exclusions_01]
```
but while less legible I guess it would be ok? (maybe that can be customized too?)
